include/simdjson/generic/ondemand/array.h

```diff
@@ -31,7 +31,18 @@ class array {
    * Part of the std::iterable interface.
    */
   simdjson_really_inline simdjson_result<array_iterator> end() noexcept;
-
+  /**
+   * This method scans the array and counts the number of elements.
+   * The count_elements method should always be called before you have begun
+   * iterating through the array: it is expected that you are pointing at
+   * the beginning of the array.
+   * The runtime complexity is linear in the size of the array. After
+   * calling this function, if successful, the array is 'rewinded' at its
+   * beginning as if it had never been accessed. If the JSON is malformed (e.g.,
+   * there is a missing comma), then an error is returned and it is no longer
+   * safe to continue.
+   */
+  simdjson_really_inline simdjson_result<size_t> count_elements() & noexcept;
```

Comment:
Checking about the `&` here: this restricts the method so that you can't do `value.get_array().count_elements()`--you have to put it in a variable first. Was that intentional?

Reply:
Can you check whether you are satisfied by this answer: https://github.com/simdjson/simdjson/pull/1577#issuecomment-843548228

If you are, I will add an explanation (in the code) for future coders.

Reply:
Ok, it doesn't seem unreasonable for this since you'll usually be doing a count and then an iteration. While it does rule out a few perfectly reasonable uses, the gain could well be worth it.

The bug you mentioned seems like it should return OUT_OF_ORDER_ITERATION, which hopefully should give people a better idea what's going on ... usually assertions are reserved for bugs in the library (though I don't remember anymore if I was really rigorous about that :). If we're asserting to users a lot, we'll probably at least want the assertions to start suggesting to people what went wrong. For the future though.

Reply:
If you look at @KordianD's example above, it 'looks' fine (if you look quickly) and seems like a bug in our code. It does not matter much what the error being thrown is... I think it is a surprising effect.

The surprise comes from the fact that when you grab something and ask for its length, you do not feel like you consumed it.